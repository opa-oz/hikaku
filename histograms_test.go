package hikaku

import (
	"github.com/gkampitakis/go-snaps/snaps"
	"testing"
)

func TestCalcHistogram(t *testing.T) {
	image, err := open("./subjects/golden.jpg")
	second, _ := open("./subjects/copper.jpg")

	histogram := calcHistogram(image, 16)
	secondHist := calcHistogram(second, 16)
	bounds := image.Bounds()

	snaps.MatchSnapshot(t, err)
	snaps.MatchSnapshot(t, histogram)

	resolution := float64(bounds.Max.X * bounds.Max.Y)

	normalizedHistogram := normalizeHistogram(histogram, resolution)
	snaps.MatchSnapshot(t, normalizedHistogram)

	secondNormalizedHistogram := normalizeHistogram(secondHist, resolution)

	red := calcHellinger(normalizedHistogram, secondNormalizedHistogram, 0)   // red
	green := calcHellinger(normalizedHistogram, secondNormalizedHistogram, 1) // green
	blue := calcHellinger(normalizedHistogram, secondNormalizedHistogram, 2)  // blue

	snaps.MatchSnapshot(t, red)
	snaps.MatchSnapshot(t, green)
	snaps.MatchSnapshot(t, blue)

	result := compareHistograms(normalizedHistogram, secondNormalizedHistogram, 16)

	snaps.MatchSnapshot(t, result)
}
