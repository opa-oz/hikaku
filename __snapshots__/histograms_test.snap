
[TestCalcHistogram - 1]
nil
---

[TestCalcHistogram - 2]
[][3]int{
    {0, 0, 0},
    {450, 670, 760},
    {3973, 12027, 9588},
    {22317, 30141, 23236},
    {42528, 35795, 40373},
    {24091, 18456, 21181},
    {13392, 14213, 9833},
    {12038, 8052, 8009},
    {8571, 7042, 6936},
    {5760, 6085, 6741},
    {4911, 5868, 7858},
    {5002, 5215, 4440},
    {5344, 4932, 5301},
    {5207, 5427, 7919},
    {2903, 2766, 3957},
    {5143, 4986, 4811},
}
---

[TestCalcHistogram - 3]
[][3]float64{
    {0, 0, 0},
    {0.0005952380952380953, 0.0008862433862433862, 0.0010052910052910052},
    {0.005255291005291005, 0.01590873015873016, 0.012682539682539682},
    {0.02951984126984127, 0.039869047619047616, 0.030735449735449734},
    {0.05625396825396825, 0.0473478835978836, 0.053403439153439154},
    {0.03186640211640212, 0.024412698412698414, 0.028017195767195767},
    {0.017714285714285714, 0.01880026455026455, 0.013006613756613756},
    {0.015923280423280423, 0.01065079365079365, 0.010593915343915345},
    {0.011337301587301587, 0.009314814814814814, 0.009174603174603174},
    {0.007619047619047619, 0.008048941798941799, 0.008916666666666666},
    {0.006496031746031746, 0.0077619047619047615, 0.010394179894179895},
    {0.006616402116402117, 0.006898148148148148, 0.005873015873015873},
    {0.007068783068783069, 0.006523809523809524, 0.007011904761904762},
    {0.006887566137566138, 0.007178571428571428, 0.010474867724867725},
    {0.00383994708994709, 0.0036587301587301586, 0.005234126984126984},
    {0.006802910052910053, 0.006595238095238095, 0.006363756613756614},
}
---

[TestCalcHistogram - 4]
float64(0.16109142855326936)
---

[TestCalcHistogram - 5]
float64(0.12159557054212825)
---

[TestCalcHistogram - 6]
float64(0.16487199138352293)
---

[TestCalcHistogram - 7]
float64(0.44755899047892056)
---
